---
title: "Mathematical modeling in ecology and evolutionary biology"
author: Amber Gigi Hoi & Zoë Humphries
output: pdf_document
---


# Lesson preamble: 

> ### Lesson objectives:
>
> - Introduce students to the world of mathematical modelling -- the art, the mechanics, the possibilities, and the limitations
> - Be comfortable with -- and even appreciate -- math and theory 
> - Equip students with tools to construct their first model 
> - Work through examples of how to use models to solve real world problems
>
> ### Lesson outline:
> Total lesson time: 3 hours
>
> - Introduction to mathematical models (10 min)
> - Learn about the different moving parts of a model and common things you can do to them and get out of them (20 min)
> - Go through examples of some classic models in EEB (30 min)
> - Working through an ecological model (1 hour)
> - Working through an example of evolution model (1 hour)

---

# Introduction


## What is a model? Why do we model? 


Up until this point in this course, we've been working with statistical models with the goal to characterize relationships between variables and their significance in existing data. Today, we are going to venture into another branch of applied mathematics and learn about _mathematical models_. A mathematical model (a.k.a. first-principle models) is a description of a system using a series of equations. These models are used extensively throughout the natural and social sciences to help explain and predict the behavior of various systems of interest. 

The world is governed by series of natural laws and "truths" known as first principles. For instance, ocean currents flow in a certain way, and some salmon species reproduce just once in their life. Mathematical models provide us with an opportunity to study the intricacies of biological systems under a framework of first principles. More accurately, models forces us to explicitly lay out our conception of "reality" and to confront those intuitiions. Intuition is perhaps the biggest pitfall in biology -- everybody loves to tell a good story where everything works out and everything makes perfect sense, but how rigorously have we actually weeded them before accepting story as dogma? When you construct a model, you have no choice but to verify every single last piece of information you wish to write. 

Technically, one can describe any biological phenonemon with math. At the same time, however, it is rarely the goal to rewrite "reality" in math. The most elegant models are those able to describe succinctly a system's most essential attributes, and the best theoreticians are those with the state-of-the-art knownledge of the system and the clarity of mind to know exactly what those important features are _given the questions being asked_. Sometimes we ask grand questions about driviers of macro system change, other times we would kill to find that devil in the details. Knowing how best to represent your system and answer your question is a skill that comes with experience (read: lots of trials and lots and _lots_ of errors).

What models are good for: 

* Guiding intuition regarding how process and mechanisms work.
* Pointing out logical flaws in an argument when the math don't make sense.
* Testing hypotheses that are otherwise untestable due to ethnical concerns and/or logistical constraints.
* Generating new testable hypotheses and predictions and can help with experimental design.
* Provide a new framework on how to think about a problem.

Limitations of a model:

* Only as interesting as the motivating question.
* Only as robust as the theory that went into building it.
* All models are wrong, some are useful.
* Without data, models are only good for telling us what is possible, but not what happened or what will happen.

Diversity promotes science and progress. Thus, any good researcher should strive to have at least a basic understanding of both the empirical and theoretical sides of science in order to facilitate collaboration and synthesis, and to push the frontiers of knowledge.


## Setup

Here are the packages we will be using today:


```{r message=FALSE, warning=FALSE}

library(tidyverse)
library(deSolve)

```


# Getting started with theory 


## Some simple terminologies


* Variable -- quantity changes over time, a.k.a. the things we are tracking.
* Parameter -- set values, quantity doesn't change over time, a.k.a. the things we know or think that we know.
* Dynamics -- fancy way of saying patterns of change over time.
* Equilibria -- the state at which the system stops changing.
* Stability -- whether the system will remain at a particular state when perturbed.
* Sensitivity -- how readily a system will change when perturbed.


## Some simple classifications


Two basic categories based on how to solve: analytic and numeric 

* A model is solved analytically when exact solutions of desired outputs are obtained. 
* Numeric methods allows you to approximate solutions when the system is too complex to solve.

Two basic categories based on type of dynamics

* Discrete time models describe change in discrete time step with recursion equations. These are useful when modelling semelparous organisms.
* Continous time models allow variables to change at any point in time, and is described with differential equations. These are useful when modelling iteroparous organisms.

Two basic categories based on nature of change: deterministic and stochastic 

* Deterministic models assumes the future is determined entirely by model mechanics. 
* Stochastic models assumes that there are random events that affect the biological processes of interest.

There is always some element of randomness in real life. Whether to model that randomness explicitly depends on how big of a role it plays in determining the dynamics of your system.


## Learning basic model mechanics through classic models in EEB


In this section, we will learn different ways to represent system change from one time step to another by examining and dericing some classic models in EEB. 


### Exponential growth

The first population model we examine is the exponential growth model, which describes a population that grows without restriction, e.g., bacteria on a petri dish.

* In discrete time: literally what is $N_{t+1}$ given $N_t$.

$$N_{t+1} = RN_t$$

* In continous time: population growth is described as change in $N$ per change in $t$. $dN/dt$ is thus the rate of population change, or the _growth rate_ of a population. 

$$\frac{dN}{dt}=rN$$

Note that $R$ and $r$ are **not** the same. $R$ is termed the _reproductive factor_ (or ratio, or number), and it represents the number of replicates each individual makes of themselves in a subsequent time step. On the other hand, $r$ describes the per capita change in number of individual from one generation to the next and is referred to as the _intrinsic growth rate_. Both of these measure reproductive output, and you can convert from one to the other with this simple formula $R=1+r$.

If we want to find out what the population size is at $t=t+1$, we simply plug the necessary parameters into the recursion equation listed like this:

```{r} 

# Set up some master controls such that it is easier to customize model later on

N0 <- 10 # Initial population size
R <- 1.2 # Reproductive factor

N1 <- R*N0; N1 

```


Most of the time though, biologists want to track the population for more than just one time step. In that case, we could either continue to do math like this:

$$N_{t+5} = R(R(R(R(RN_0))))$$

which simplifies to 

$$N_{t+5} = R^5 (N_0)$$

and then further simplifies to

$$N_{t+n} = R^n(N_0)$$

We coould also take a graphical approach and draw out the dynamics of the system for 50 time steps.


```{r}

N0 <- 10 # Initial population size
R <- 1.2 # Reproductive ratio
times <- 50 # Number of time steps

N <- c() # Make empty vector to store population size at each time step
N[1] <- N0 # Initiate vector

# Make for loop to iterate through the model 50 times
# Running (times-1) iterations gets us to generation N[i+1]

for (i in 1:(times-1)) { 
  N[i+1] = R*N[i]
  }

# Plot the newly loaded N vector 

plot(seq(1, times, 1), N)

```


As expected, this population is literally spinning out of control! In reality, however, even bacteria on a petri dish can't grow forever. What are some factors that may limit population growth?


### Logistic growth

Organisms often find themselves limited by resource availability as their numbers grow. This causes their per capita growth rate to decreases as the population increases towards the resource limit known as the _carrying capacity_, $K$.

* In discrete time: 

$$N_{t+1} = R(\frac{K-N_t}{K})N_t$$

* In continous time: 

$$\frac{dN}{dt} = r(\frac{K-N}{K})N$$

When interpreted in plain English, the growth rate of the population is being discounted depending on how close to $K$ it is. When the population is very small, $\frac{K-N}{K} \approx 1$ and logistic growth approximates exponential growth. When the population is close to $K$, $\frac{K-N}{K} \approx 0$ and thus the population size stops increasing. 

_Challenge: simulate a logistic growth model with the parameters provided_


```{r}

N0 <- 10 # Initial population size
r <- 0.5 # Intrinsic growth rate
K <- 1000 # Carrying capacity
times <- 50 # Number of time steps

```
```{r eval=FALSE, echo=FALSE}

N <- c() # Make empty vector to store population size at each time step
N[1] <- N0 # Initiate vector

for (i in 1:(times-1)) { 
  N[i+1] = N[i]+r*N[i]*(1-(N[i]/K))
  }

plot(seq(1, times, 1), N)

```


### When species are in competition -- the Lotka-Volterra model

This example will show case how to model two interacting populations, specifically, populations in competition for a limiting resource. Focusing on a continous time example, we will first write down the logistic growth models of two populations (differentiated by subscripts), and then add in the effect of their interaction.

$$\frac{dN_1}{dt} = r(\frac{K-N_1}{K})N_1$$
$$\frac{dN_2}{dt} = r(\frac{K-N_2}{K})N_2$$

What does competition do to organisms? Well, many things, but in terms of population dynamics, competition hinders the growth of a population, and the degree of reduction in growth is governed by the _competition coefficient_, $\alpha_{ij}$. The subscripts $i$ and $j$ correspond to our two populations, and the order of the subscript tells us who is affecting whom. For example, $\alpha_{ij}$ represents the effect of $j$ on $i$, whereas $\alpha_{ji}$ represent the effect of $i$ on $j$.

Adding these terms into the population growth equations gives the classic Lotka-Volterra model for competition: population growth is governed jointly by how close it is to carrying capacity, but also the effect of the competing species has on it (strength of effect $\times$ number of individuals around to exert this effect).

$$\frac{dN_1}{dt} = r(\frac{K- N_1 - \alpha_{12} N_2}{K})N_1$$
$$\frac{dN_2}{dt} = r(\frac{K - N_2 - \alpha_{21} N_1}{K})N_2$$


### Models of evolution 

A population genetics model which describes change in allele frequency in a population over time. In this example, we will work with a one-locus haploid model for simplicity.

Consider a population of individuals each carrying one of two variants of a certain gene, let's call them alleles _A_ and _a_ because we are creative like that. Similar to ecologists, evolutionary biologists are often interested in the relative survival abilities (denoted $W$) of individuals with different genotypes. The dynamics of these two sub-populations can be expressed simply as

$$N_{A, t+1} = W_A N_{A, t}$$

$$N_{a, t+1} = W_a N_{a, t}$$

Typically, evolutionary biologists are concerned with the _proportion_ of a certain allele in a population rather than changes to their absolute numbers. Here, we will use $p$ and $q$ to denote the proportions of the _A_ and _a_ alleles in the population, respectively. Doing so allows us to leverage the fact that the proportions $p$ and $q$ must add up to 1, and therefore the dynamics of the allele frequencies in a population can be described with just one equation:

$$p_{t+1}=\frac{N_{A,t+1}}{N_{A,t+1}+ N_{a,t+1}}$$

$$p_{t+1}=\frac{W_A N_{A,t}}{W_A N_{A,t}+W_a N_{a,t}}$$

In order to write an equation that tracks change in allele frequency over time, as promised, we need to transform $N$ (the number of each type) to $p$ (proportion). Through the magic of algebra, we have

$$p_{t+1}=\frac{W_A p_t}{W_A p_t+W_a q_t}$$

Note that $N{a}$ is **not** interchangeable with $p$, the algebra simply worked out this way.

Lastly, remember how $p+q=1$? 

$$p_{t+1}=\frac{W_A p_t}{W_A p_t+W_a (1-p_t)}$$


## How to construct a model


### Step 0: Breathe

Don't panic, and try not to be overwhelmed. Trust us when we say it is not as daunting as it appears. This doesn't mean that it is easy to write a model either, but there are lots of resources available to help you acheieve your ambition. The biggest hurdle is often to simply _start_. Know that whatever you do, these are just numbers and symbols and words on paper and on the screen, you can't break anything or hurt anything (except maybe R, but just turn it on and off again and all is forgiven, life will move on). So don't be afraid to start typing and playing around with ideas. Know also that you will be making mistakes and will have to troubleshoot your way through a theory project (any project really...). We all go through that, even the most experienced theoreticians make mistakes. Experience comes with time and practice, and you will learn how to quickly spot mistakes, and how to
effectively troubleshoot. None of that will happen though, if you don't start.

So let's start!

Below, we'll break down some _typical_ steps of constructing a model.

### Step 1: Formulate the question.

* What do you want to know? Start with the simplest, biologicallyreasonable description of the problem. 
* Having clear objectives is crucial!

### Step 2: Determine the basic ingredients of the model.

* Decide on which type of model to use (discrete vs. continous time, stochastic vs. deterministic, etc.).
* Define the variables and parameters you want to keep track of.
* **Think about the assumptions you are making about your system**. Given that we can't ever full replicate reality (and we don't want to anyway), we are neccessarily making simplifying assumptions about our system as we write our model. What are those assumptions? You need to make them clear to both yourself and your audience. Your model can only be interpreted in the context of thee assumptions within. 

### Step 3: Qualitatively describe the system.

* Use life-cycle diagrams and flow charts to decribe changes to the individuals/populations/communities that you're modelling.
* Make a table or list of variables and how they will interact with each other.
* Think about what are the possible outcomes and what types of output you want from your model.

### Step 4: Quantitatively describe the system.

* Using the diagrams and tables etc as guides, draft out equations that describe your system.
* Think about the functional forms of interactions in your model and how to parameterize  model (i.e., where are you going to get data?).
* Perform checks to make sure the structure of the model is logically sound (e.g., will your model return negative population size?).

### Step 5: Analyze your model

* There are abundant options regarding how to analyze a model. These range from finding equiliria and performing a stability analysis analytically, to numerically simulating population dynamics over time.
* Choose the appropriate analysis for your system and question, and make sure that the output does address the question adequately! 

### Step 6: Checks and balances 

* Check results against data or known cases.
* Consider alternatives to a simplified model -- how can this model be extended?

### Step 7: Relate results back to the question

* Interpret the results and describe any new insights obtained in the context of assumptions made.
* How general are your results? Are the results counterintuitive? Why?

We will be following this recipe (roughly) in our build-a-model workshops. 


# Built-a-model workshop 1 - Malaria transmission dynamics 


First, take a _deeeeeeeeeeeeeeep_ breath.

In this workshop, we are going all the way back to the dawn of mathematical models in biology. We will borrow from the classic malaria model orginally developed by Ross (1908) and later solidified by Macdonald (1956), and investigate the required efficacy of mosquito management programs in order to eradicate malaria. 

_Note:_ It is actually very common practice to adapt old models to answer new questions. In fact, practically _all_ of the malaria models that exist today is a variation of the Ross-Macdonald model. 

The Ross-Macdonald model is a continuous-time model which tracks the number of **s**usceptible, **i**nfected, and **r**ecovered subjects over the course of a malaria outbreak. These models are sometimes referred to as _SIR_ models, and they have been adapted to model numerous diseases beyond malaria.

Figure 1 shows a flow diagram of the malaria transmission cycle we will modelling today.

![Flow diagram illustrating a basic epidemiological model of malaria transmission. The blocks represent different host and vector classes within a population. Arrows describe possible movements between host/vector categories. Subscripts _h_ and _m_ represent hosts and mosquitoes, respectively.](image/model.png)

The principle epidemiological compartments (i.e., variables) of interest are the populations of infectious hosts ($I_h$) and infectious vectors ($I_m$). These compartments are interdependent via the hosts’ and the vectors’ rate of exposure to the parasite, given by $\beta I_m$ and $\beta I_h$, respectively, where $\beta$ represent the per capita biting rate of vectors. Upon contact, susceptible hosts ($S_h$) and mosquitoes ($S_m$) become infected with probabilities $\varphi_h$ and $\varphi_m$, respectively. For simplicity, we will assume that hosts cannot acquire immunity and so the "recovered" class was not included in this model. Instead, hosts either die of infection at rate $\alpha$ or recover and return to $S$ class at rate $\gamma$. Vectors were assumed to not be affected by the parasite and remain infected for life (i.e., no $\alpha$ and $\gamma$ for vectors). All host and vector die of natural causes at rate $\mu_h$ and $\mu_m$, respectively. The host and vector populations were assumed to be closed and constant over time such that all deaths are immediately replaced by immigration of individuals into susceptible classes, represented by the symbol $\tau$.  

After all that prefacing, we're finally ready to do math! 

Recall that, in contrast to recursion equations where we describe directly the change in the value of a variable, an ordinary differential equation (ODE) model describes _how_ the population is changing. We will follow this simple formula when we translate our flow diagram to our equations:

$$\frac{dN}{dt} = rate\;of\;increase - rate\;of\;decrease \\
= flow\;rates\;along\;arrows\;entering\;the\;compartment - flow\;rates\;along\;arrows\;exiting\;the\;compartment$$

Note that the order of the entering and exiting of compartments is irrelevant because each time step is so small, it doesn't matter which event happens first. 

The resulting model is thus made up of a system of four ODEs, two for hosts and two for vectors, written as follows:

$$\frac{dS_h}{dt} = \tau_t - (\mu_h +\beta I_m \varphi_h)S_h + \gamma I_h$$

$$\frac{dI_h}{dt} = (\beta I_m \varphi_h)S_h - (\gamma + \mu_h + \alpha_h) I_h$$

$$\frac{dS_m}{dt} = \tau_m - (\mu_m + \beta I_h \varphi_m)S_m$$

$$\frac{dI_m}{dt} = (\beta I_h \varphi_m)S_m - \mu_m I_m$$

Now that we have our model written out, it is R's turn to do some math for us. Specifically, R will _solve_ this system for us and save us from a lot of headache and paper. To solve a system of ODEs means we will integrate the equations to find expressions that describes the state of each population as a function of the remaining variables and parameters.

We will code model like we did before. First, we will set up a bunch of master controls for easy model manipulation. Then, we will write out our ODE model as a function. Lastly, we will call the ODE solver _lsoda()_ to put all this together. 

### Define required parameters for the main program

* Time vector


```{r}

times <- 50; # Time is per day               
timevec <- seq(0, times, 1) # Vector of times for which integration is evaluated 

```


* Assign values for model parameters. These parameters are mostly personal guesstimates. 


```{r} 

# Units are assumed to be /day

mu.h <- 0.05; mu.m <- 0.1 # Natural mortality
psi.h <- 0.1; psi.m <- 0.1 # Probability of infection upon exposure
b <- 0.25 # Biting rate
N <- 500 # Total population size (closed population)
beta <- b/N # per capita biting rate
gamma <- 0.03 # Recovery rate (host only)
alpha <- 0.2 # disease mortality rate (host only)

# String it all up as a vector

parms <- c(mu.h, mu.m, psi.h, psi.m, b, N, beta, gamma, alpha)

```


* Give the model something to get it going a.k.a. assign initial conditions


```{r}

s.h0 <- 450; i.h0 <- 50 # Host
s.m0 <- 200; i.m0 <- 50 # Vector

# Make into another vector

Y0 <- c(s.h0, i.h0, s.m0, i.m0)

```


* Define the function that specifies the ODE model called by _lsoda()_ a.k.a. type out equations.


```{r}

odeequations <- function(t,y,parms) 
{
  # Retype parms and initial conditions
  # This will link parameters set "outside" of the function to the equations "inside"

  # Natural mortality parms
  mu.h<-parms[1]; mu.m<-parms[2]
  
  # Probability of infection parms
  psi.h<-parms[3]; psi.m<-parms[4] 
  
  # Biting rate parms
  b<-parms[5]; N<-parms[6]; beta<-parms[7]
  
  #recovery parms
  gamma<-parms[8];
  
  #virulence 
  alpha<-parms[9];
  
  # Initial conditions
  s.h <- y[1]; i.h <- y[2]; s.m <- y[3]; i.m <- y[4]
  
  # These are the differential equations
  dsh = (mu.h*s.h + (mu.h + alpha)*i.h) - (mu.h + beta*i.m*psi.h)*s.h + gamma*i.h
  dih = beta*i.m*psi.h*s.h - (gamma + mu.h + alpha)*i.h
  dsm = mu.m*(s.m + i.m) - (mu.m + beta*i.h*psi.m)*s.m 
  dim = beta*i.h*psi.m*s.m - mu.m*i.m 
  
  return(list(c(dsh, dih, dsm, dim))); 
  
}

```


* Call ode-solver _lsoda()_ to integrate ODEs 


```{r}

odeoutput1 <- lsoda(Y0,timevec,odeequations,parms)

```


Anddddd that's it! We are done! We can now look at our shiny model outputs.


```{r}

head(odeoutput1)

```


The output dataframe consist of five columns, one for time, and the next four represent the population size at each of those time points. The columns are organized in the same order we typed out our equations. Therefore, column 1 represent susceptible hosts, column 2 infected hosts, etc. To make our lives easier from now on, we'll go ahead and rename these columns. We're also going to force the output into data frame.


```{r}

colnames(odeoutput1)[2:5] <- c("S.h", "I.h", "S.m", "I.m")
odeoutput1 <- as.data.frame(odeoutput1)

```


We can inspect the dynamics of this system by ploting out these trajetories. 


```{r}

ggplot() +
  geom_line(data=odeoutput1, aes(x=time, y=S.h), colour="black") +
  geom_line(data=odeoutput1, aes(x=time, y=I.h), colour="blue") +
  labs(y="Hosts") + 
  theme_classic()

ggplot() +
  geom_line(data=odeoutput1, aes(x=time, y=S.m), colour="black") +
  geom_line(data=odeoutput1, aes(x=time, y=I.m), colour="blue") +
  labs(y="Hosts") + 
  theme_classic()
  
```

